### ISSUE TITLE
Ticket custom fields with attachments

### ISSUE THREAD

**Issue by lucasabond** [DaveTCode/FreshdeskApiDotnet#200]
The custom fields are setup as a Dictionary, which works fine with JSON serialization, but it does not work properly with the multipart form serializer. I ran into this when trying to create a ticket that includes attachments and custom fields.

The easiest solution seems to be changing the custom fields attribute to a simple object class. I create a custom fields class in my project with the custom fields I need and the serialization works properly with and without attachments. 

`public Dictionary<string, object>? CustomFields { get; }`

Becomes

`public object? CustomFields { get; }`

And then the custom fields class in my project looks something like this:

```csharp
public class TicketCustomFields {
        [JsonProperty("cf_customfield1")]
        public int CustomField1{ get; set; }

        [JsonProperty("cf_customfield2")]
        public string CustomField2{ get; set; }
    }
```

This would require a little documentation, but I am happy to submit a pull request if this change makes sense.

**Comment by trejjam** [comment]
Hi,
We are using https://github.com/tiberhealth/MultipartSerializer for multipart serialization. It will be nice if you can find out how is dictionary serialized there and how to configure the library to serialize it properly. So users of this library will not be forced to implement their own classes. Also having just an object type is not great (if the data are not a random object). If it's not possible this way then I am open to relax the type.

If a plain dictionary is not enough we can create a extending dictionary with special handling for serialization, or something similar. The idea is to not force people to think (when you have a dictionary you know it has keys and values, with an object you do not know what to do)

**Comment by lucasabond** [comment]
Hi,

Yes, I agree that it is not ideal. And it would also be a breaking change for anyone who upgraded the library. I assume some people may be using the custom fields without the multi part serialization, which works fine as is.

I did look at the way the multipart serializer is working. Currently it renders the custom fields as (from my memory)

```
Content-Disposition: form-data; name="custom_fields[]"

[customFieldName: customFieldValue]
```

And what is needed for FD is

```
Content-Disposition: form-data; name="custom_fields[customFieldName]"

customFieldValue
```

I then just jumped over to a custom class, seeing that is was a basic array-style serialization that was expected. 

I did not dig into the tiberhealth project, but if there is a way to implement a custom serialization option, maybe just a MultipartArray attribute, that is probably the way to go.

**Comment by binon** [comment]
Any updates on this please? How do I approach it? I need to send both custom fields and attachments at the same time and having issues

![image](https://github.com/DaveTCode/FreshdeskApiDotnet/assets/18263516/3c443c0a-4d45-4338-8f01-528ba12a7f73)

**Comment by trejjam** [comment]
Hi, it is in the state where Lucas left it. I am open to supervising a PR that solves this issue.

As I wrote in the previous comment, I prefer to use `MultipartSerializer` library that is already present in the project and configure it properly for custom field serialization. For that, someone needs to investigate if it's possible, and how to do it.

**Comment by binon** [comment]
Thanks Jan, I'll see what I can do.

**Comment by IanCodeAces** [comment]
Hello all

I've ran into a similar issue and I think I know what the problem is, (I've only spent an hour or so looking at this so may have missed something)

So what we are currently sending in our multipart form data is something like this

```
--BoundaryString
Content-Disposition: form-data; name="custom_fields[]"
Content-Type: text/plain

[{"keyname","keyvalue"},{},{}]
```

But Freshdesk doesn't want an array when sending this as multipart form data - hence why you get the response ""Value set is of type Array.It should be a/an key/value pair"

What we actually need to send is something that looks like this:

```
--BoundaryString
Content-Disposition: form-data; name="custom_fields[custom_field_key1]"
Content-Type: text/plain

custom_field_value1
```
 If you were trying to send it as form data in Postman it would look like this:
  
<img width="350" alt="image" src="https://github.com/DaveTCode/FreshdeskApiDotnet/assets/113179156/1a6cf121-ac56-4c2f-9113-0cb1b0295552">

I've looked at the Serializer library currently being used and I can't see anything on formatting that could be changed to give us this output - not without changing their library as well. The rest of the library works as expected.

I might have a workaround that needs a little bit of experimentation but we'd effectively have to map custom fields manually

I'll let you know how I get on

**Comment by IanCodeAces** [comment]
An update

So i've got this working - it's not the prettiest and when I get a bit more time I can refactor some of it and perhaps even look at creating a custom MultipartFormData serializer that will handle everything. I submitted a PR

Essentially this is a workaround to allow users to create new tickets that contains both custom fields and attachments. The MultipartSerializer used in the project cannot handle the custom fields and produces output that isn't consistent with what Freshdesk expects to receive. As a workaround, we strip out custom fields from the original payload and serialize ticket with attachments without CF and then manually add them back into the MultiPartFormDataContent we are going to send to Freshdesk

Changes:

1. Added new model for CreateTicketRequestWithoutCustomFields (mapped from original request and processed by serializer for attachments)

2. Added static HttpRequestMessageExtensions to seperate logic for building the message

2. Updated FreshdeskHttpClient to use new extensions and produce output

I've tested this on our instance and it's working - we have an Azure Function App that acts as a webhook for a very specific form on a WordPress website and it's pumping form requests into Freshdesk
